Mathlib/FieldTheory/PurelyInseparable/AdjoinPthRoots.lean

```diff
@@ -0,0 +1,130 @@
+  inseparable
+
+- `adjoin_pth_roots.mem_frobenius_img`: every element of `S` is in the image of the frobenius
+  morphism on `adjoin_pth_roots p S`.
+
+-/
+
+@[expose] public section
+
+open IntermediateField
+
+noncomputable section
+
+variable {k : Type*} [Field k]
+variable (p : ℕ) [ExpChar k p]
+
+instance : ExpChar (AlgebraicClosure k) p := ExpChar.of_injective_algebraMap' k _
+
+/-- Given a field `k` of exponential characteristic `p` and a subset `S` of `k`, the field
+`adjoin_pth_roots p S` is obtained by adjoining all `p`-th roots of elements of `S` to `k`. We
+construct this as an object of type `IntermediateField k (AlgebraicClosure k)`, which automatically
+gives us the structure of a field and of a `k`-algebra. -/
+def IntermediateField.adjoinPthRoots (S : Set k) : IntermediateField k (AlgebraicClosure k) :=
+  IntermediateField.adjoin k <|
+    (frobenius (AlgebraicClosure k) p) ⁻¹' ((algebraMap k (AlgebraicClosure k)) '' S)
+
+lemma adjoinPthRoots_mono {S T : Set k} (hST : S ⊆ T) :
+    adjoinPthRoots p S ≤ adjoinPthRoots p T :=
+  IntermediateField.adjoin.mono k _ _ <| Set.preimage_mono <| Set.image_mono hST
+
+/-- If the set `S` whose `p`-th roots we adjoin is finite, then the obtained field extension
+`(adjoin_pth_roots p S) / k` is finite. -/
+lemma adjoinPthRoots_finite_of_finite (S : Set k) [Finite S] :
+    FiniteDimensional k (adjoinPthRoots p S) := by
+  -- The set of elements to adjoin to `k` is finite:
+  have hFin : Finite ((frobenius (AlgebraicClosure k) p) ⁻¹'
+      ((algebraMap k (AlgebraicClosure k)) '' S)) := by
+    have hFin' : Finite ((algebraMap k (AlgebraicClosure k)) '' S) := by infer_instance
+    exact Set.Finite.preimage (Set.injOn_of_injective (frobenius_inj _ _)) (hFin')
+  apply IntermediateField.finiteDimensional_adjoin
+  -- It remains to show that every element of the set
+  -- `` ((frobenius (AlgebraicClosure k) p) ⁻¹' ((algebraMap k (AlgebraicClosure k)) '' S))``
+  -- is integral over `k`.
+  intro s hs
+  apply IsIntegral.of_pow (n := p) (expChar_pos k p)
+  have hs_mem : s ^ p ∈ (algebraMap k (AlgebraicClosure k))'' S := by
+    simp_all only [Set.mem_preimage, Set.mem_image]
+    obtain ⟨w, hw⟩ := hs
+    use w
+    rw [hw.2]
+    exact ⟨hw.1, rfl⟩
+  rw [Set.mem_image] at hs_mem
+  obtain ⟨y, hy⟩ := hs_mem
+  rw [← hy.2]
+  exact isIntegral_algebraMap
+
+/-- The field extension `(adjoin_pth_roots p S) / k` is purely inseparable. -/
+instance adjoinPthRoots_purelyInseparable (S : Set k) :
+    IsPurelyInseparable k (adjoinPthRoots p S) := by
+  unfold adjoinPthRoots
+  rw [IntermediateField.isPurelyInseparable_adjoin_iff_pow_mem k (AlgebraicClosure k) p]
+  intro s hs
+  use 1
+  simp_all only [Set.mem_preimage, Set.mem_image, pow_one, RingHom.mem_range]
+  obtain ⟨w, h1, h2⟩ := hs
+  use w
+  rw [h2]
+  rfl
+
+/-- If `y ∈ S`, then there is an element `x ∈ adjoin_pth_roots p S` with the property that
+`y = x ^ p`. -/
+lemma adjoinPthRoots_mem_frobenius_img {S : Set k} {y : k} (hy : y ∈ S) :
```

Comment:
```suggestion
lemma mem_map_frobenius_adjoinPthRoots {S : Set k} {y : k} (hy : y ∈ S) :
```
It might be useful to also read the [naming convention](https://leanprover-community.github.io/contribute/naming.html).